Plugins/NativeCamera/Source/NativeCameraAndroid.cpp

```diff
@@ -0,0 +1,404 @@
+                if (eglMakeCurrent(display, 0/*surface*/, 0/*surface*/, context) == EGL_FALSE)
+                {
+                    throw std::runtime_error{"Unable to create a shared GL context for camera texture."};
+                }
+            }
+
+            cameraOESTextureId = GenerateOESTexture();
+            glGenTextures(1, &cameraRGBATextureId);
+            glBindTexture(GL_TEXTURE_2D, cameraRGBATextureId);
+            glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, this->width, this->height, 0, GL_RGBA, GL_UNSIGNED_BYTE, nullptr);
+            glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
+            glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
+            glGenerateMipmap(GL_TEXTURE_2D);
+
+            glBindTexture(GL_TEXTURE_2D, 0);
+
+            glGenFramebuffers(1, &frameBufferId);
+            glBindFramebuffer(GL_FRAMEBUFFER, frameBufferId);
+            glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D, cameraRGBATextureId, 0);
+
+            glBindFramebuffer(GL_FRAMEBUFFER, 0);
+
+            cameraShaderProgramId = android::OpenGLHelpers::CreateShaderProgram(CAMERA_VERT_SHADER, CAMERA_FRAG_SHADER);
+
+            // Create the surface and surface texture that will receive the camera preview
+            surfaceTexture.initWithTexture(cameraOESTextureId);
+            surface.initWithSurfaceTexture(surfaceTexture);
+
+            // open the front or back camera
+            cameraManager = ACameraManager_create();
+            auto id = getCamId(frontCamera);
+            ACameraManager_openCamera(cameraManager, id.c_str(), &cameraDeviceCallbacks, &cameraDevice);
+
+            textureWindow = surface.getNativeWindow();
+
+            // Prepare request for texture target
+            ACameraDevice_createCaptureRequest(cameraDevice, TEMPLATE_PREVIEW, &request);
+
+            // Prepare outputs for session
+            ACaptureSessionOutput_create(textureWindow, &textureOutput);
+            ACaptureSessionOutputContainer_create(&outputs);
+            ACaptureSessionOutputContainer_add(outputs, textureOutput);
+
+            // Prepare target surface
+            ANativeWindow_acquire(textureWindow);
+            ACameraOutputTarget_create(textureWindow, &textureTarget);
+            ACaptureRequest_addTarget(request, textureTarget);
+
+            // Create the session
+            ACameraDevice_createCaptureSession(cameraDevice, outputs, &sessionStateCallbacks, &textureSession);
+
+            // Start capturing continuously
+            ACameraCaptureSession_setRepeatingRequest(textureSession, &captureCallbacks, 1, &request, nullptr);
+
+            if (eglMakeCurrent(display, 0/*surface*/, 0/*surface*/, currentContext) == EGL_FALSE)
+            {
+                throw std::runtime_error{"Unable to restore GL context for camera texture init."};
+            }
+        });
+    }
+
+    CameraInterfaceAndroid::~CameraInterfaceAndroid()
```

Comment:
Bad things will happen if the destructor is called while async construction is still in flight, right? Should we do something to defend against that?